### DISCUSSION TITLE
Google Cloud Storage provider environment variable support

### DISCUSSION THREAD

**Discussion by stx-chris** [directus/directus#7257] [Feature Requests]
Our use case prohibits exposing the GCS credentials in a key file and referencing it in the `STORAGE_<LOCATION>_KEY_FILENAME` variable. I would suggest to extend the parsing routine to allow for variables as mentioned in the [StorageOptions](https://googleapis.dev/nodejs/storage/latest/global.html#StorageOptions).

For some reason Google decided to mix camelCase and underscores in their Cloud Storage constructor, see https://googleapis.dev/nodejs/storage/latest/global.html#StorageOptions (`projectId` and `keyFilename`, but `credentials.client_email` and `credentials.private_key`). Since Directus converts the storage config variables into camelCase only (`/api/src/utils/get-config-from-env.ts`), any intended underscores are lost. 

Solving this issue without introducing a breaking change might be structured as follows:

1. Encode an intended underscore by three underscores "___" in the key, eg. `STORAGE_GCS_CREDENTIALS__CLIENT___EMAIL`
2. Adjust the replacement routine to prevent this pattern from being changed:
```
                 if (key.includes('__')) {
			  const path = key
                                  .split('___')
                                  .join('@')  // <- temporarily replace intended underscores by an invalid character
                                  .split('__')
				.map((key, index) => (index === 0 ? transform(transform(key.slice(prefix.length))) : transform(key)));
			set(config, path.join('.'), value);
		} else {
			config[transform(key.slice(prefix.length))] = value;
		}
```
3. Make sure the underscores are restored after the transform: 
```
                if (type === 'camelcase') {
			return camelcase(key).split('@').join('_');
		} else if (type === 'underscore') {
```

This results in the desired `config` object:

```
{
  driver: 'gcs',
  root: '/',
  projectId: '---',
  credentials: {
    private_key: '...'
    client_email: '...'
  },
}
```

Happy to provide a PR.

P.S. The double execution of `transform` in the `map()` seems to be redundant and affects the proposed solution in case there is no double underscore "__" in the key name. Therefore I would suggest to remove it.

**Comment by rijkvanzanten** [comment]
Or, we could support custom JSON structures in the env var as an escape hatch for these edge cases, which I actually implemented just 2 days ago 🙂 https://github.com/directus/directus/pull/7201

`STORAGE_GCS_CREDENTIALS='{ "private_key": "abc", "client_email": "def" }'`

**Reply by stx-chris** [reply]
well...in that case 🤠 thx!

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
